### DISCUSSION TITLE
Getting `Expected type expression but received "UnionType"` from pyright when using model as type with v0.0.14

### DISCUSSION THREAD

**Discussion by ukalwa** [fastapi/sqlmodel#737] [Questions]
### First Check

- [X] I added a very descriptive title here.
- [X] I used the GitHub search to find a similar question and didn't find it.
- [X] I searched the SQLModel documentation, with the integrated search.
- [X] I already searched in Google "How to X in SQLModel" and didn't find any information.
- [X] I already read and followed all the tutorial in the docs and didn't find an answer.
- [X] I already checked if it is not related to SQLModel but to [Pydantic](https://github.com/samuelcolvin/pydantic).
- [X] I already checked if it is not related to SQLModel but to [SQLAlchemy](https://github.com/sqlalchemy/sqlalchemy).

### Commit to Help

- [X] I commit to help with one of those options 👆

### Example Code

```python
from typing import Optional

from sqlmodel import Field, Session, SQLModel, create_engine


class Hero(SQLModel, table=True):
    id: Optional[int] = Field(default=None, primary_key=True)
    name: str
    secret_name: str
    age: Optional[int] = None


hero_1 = Hero(name="Deadpond", secret_name="Dive Wilson")

engine = create_engine("sqlite:///database.db")


SQLModel.metadata.create_all(engine)

with Session(engine) as session:
    session.add(hero_1)
    session.commit()
    session.refresh(hero_1)
    print(hero_1)

def get_hero(user_id: int) -> Hero | None: # Pyright error here: Expected type expression but received "UnionType" 
    with Session(engine) as session:
        return session.exec(select(Hero).where(cls.id == user_id)).first()
```

### Description

The issue seems to go away when I change the inheritance order in SQLModelMetaclass here

https://github.com/tiangolo/sqlmodel/blob/b892504141c842975b7a30156ab21755719e1d31/sqlmodel/main.py#L389-L390
to

```python
class SQLModelMetaclass(DeclarativeMeta, ModelMetaclass): 
    __sqlmodel_relationships__: Dict[str, RelationshipInfo]
```

### Operating System

macOS

### SQLModel Version

0.0.14

### Python Version

3.10.11

**Comment by Philipp258** [comment]
I have the same issue in my codebase, for me it worked to use `Optional[Model]` instead of `Model | None`

**Reply by 2jun0** [reply]
It works. Thank you

**Reply by mbertrand-kintsugi** [reply]
While this works, this is not technically correct. The `|` has been supported as a Union since Python 3.10. In our code, we _accept_ either a `None` value or a "class instance" value. We don't provide a default, therefore `Optional` isn't correct. Instead we were forced to use `Union[object, None]` which is a lot more verbose to accomplish what already worked with `object | None`.

**Reply by 9hgg** [reply]
@mbertrand-kintsugi , while I understand your point, Optional is equivalent to object | None. 
Please see the snippet from the typing lib : 
<img width="1039" alt="image" src="https://github.com/tiangolo/sqlmodel/assets/20915631/31a31f7e-fb59-48fd-8b96-2ca9d82cf858">

**Comment by YuriiMotov** [comment]
Seems that it was fixed.
Can't reproduce the error with Pyright enabled and configured as:
```
{
    "python.analysis.typeCheckingMode": "standard"
}
```
Can somebody check?
